common/junit-platform-native/src/main/java/org/graalvm/junit/platform/config/util/AnnotationUtils.java

```diff
@@ -0,0 +1,121 @@
+ *
+ * The above copyright notice and either this complete permission notice or at a
+ * minimum a reference to the UPL must be included in all copies or substantial
+ * portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.graalvm.junit.platform.config.util;
+
+import org.graalvm.junit.platform.config.core.NativeImageConfiguration;
+import org.graalvm.util.GuardedAnnotationAccess;
+import org.junit.platform.commons.support.AnnotationSupport;
+
+import java.lang.annotation.Annotation;
+import java.lang.annotation.Repeatable;
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Method;
+import java.lang.reflect.Parameter;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+public class AnnotationUtils {
+
+    public static <A extends Annotation> void forEachAnnotationOnClassMembers(Class<?> clazz, Class<A> annotationType, Consumer<A> consumer) {
+        for (A annotation : getAnnotations(clazz, annotationType)) {
+            consumer.accept(annotation);
+        }
+
+        forEachMethod(clazz, method -> {
+            for (A annotation : getAnnotations(method, annotationType)) {
+                consumer.accept(annotation);
+            }
+        });
+    }
+
+    public static <A extends Annotation> void forEachAnnotatedMethod(Class<?> clazz, Class<A> annotationType, BiConsumer<Method, A> consumer) {
+        forEachMethod(clazz, method -> {
+            for (A annotation : getAnnotations(method, annotationType)) {
+                consumer.accept(method, annotation);
+            }
+        });
+    }
+
+    public static <A extends Annotation> void forEachAnnotatedMethodParameter(Class<?> clazz, Class<A> annotationType, Consumer<A> consumer) {
+        forEachMethod(clazz, method -> {
+            for (Parameter parameter : method.getParameters()) {
+                for (A annotation: getAnnotations(parameter, annotationType)) {
+                    consumer.accept(annotation);
+                }
+            }
+        });
+    }
+
+    private static void forEachMethod(Class<?> clazz, Consumer<Method> consumer) {
+        Set<Method> methods = new HashSet<>();
+        methods.addAll(Arrays.asList(clazz.getDeclaredMethods()));
+        methods.addAll(Arrays.asList(clazz.getMethods()));
```

Comment:
Please see related comment about `getDeclaredMethods()` and `getMethods()` in class hierarchies.

Reply:
I agree, we should also only use `getDeclared<member>` methods here and traverse the class hierarchy.